### PR TITLE
create hostname update service

### PR DESCRIPTION
**Pull request by wlan0**
@alena1108 @ibuildthecloud

### CODE REVIEW

**Comment by wlan0** [comment]
https://github.com/rancher/rancher/issues/4384

**Comment by alena1108** [comment]
@wlan0 could you link to a corresponding k8s PR? the one that launches the service. You can use a fake release number for now in the download link

File: main.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	app := cli.NewApp()
+	app.Author = "Rancher Labs"
+	app.EnableBashCompletion = true
+	app.Version = VERSION
+	app.Usage = `Kubernetes Hostname Service
```
Comment (alena1108) [inline main.go]: As this service is pretty generic, lets remove k8s references from the comments

File: main.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		hostsMap[host.Hostname] = host.AgentIP
+	}
+
+	// if a host was deleted, this will be true
+	if len(rancherHosts) != len(hostsMap) {
+		changed = true
+	}
+
+	if !changed {
+		log.Info("No change detected in rancher hosts")
```
Comment (alena1108) [inline main.go]: lets log this in debug mode

**Comment by wlan0** [comment]
Has to be merged before - https://github.com/rancher/kubernetes/pull/26

File: main.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	metadataClient, err := metadata.NewClientAndWait(metadataURL)
+	if err != nil {
+		return err
+	}
+
+	interval := c.Int("update-interval")
+	rancherHosts := map[string]string{}
+
+	for {
+		err = repeat(metadataClient, rancherHosts)
```
Comment (ibuildthecloud) [inline main.go]: Please use  `metadata.Client.OnChange()` for polling.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+			// If the current host is not a part of the
+			// previous set of rancher hosts, then a new host
+			// was added
+			changed = true
+		}
+		hostsMap[host.Hostname] = host.AgentIP
+	}
+
+	// if a host was deleted, this will be true
+	if len(rancherHosts) != len(hostsMap) {
```
Comment (ibuildthecloud) [inline main.go]: Checking size doesn't seem sufficient.  Can you do proper full diff to see if the data changed.

**Comment by ibuildthecloud** [comment]
@wlan0 Can you add some tests?

File: main.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	for k, v := range hostsMap {
+		rancherHosts[k] = v
+	}
+
+	hostsWorkingFile := "/etc/hosts.backup"
+
+	if _, err := os.Stat(hostsWorkingFile); os.IsNotExist(err) {
+		// This is done to maintain a copy of the original contents
+		// of the hosts file
+		fd, err := os.Open("/etc/hosts")
```
Comment (ibuildthecloud) [inline main.go]: Just use `ioutil.ReadFile`

File: main.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+		delete(rancherHosts, k)
+	}
+
+	for k, v := range hostsMap {
+		rancherHosts[k] = v
+	}
+
+	hostsWorkingFile := "/etc/hosts.backup"
+
+	if _, err := os.Stat(hostsWorkingFile); os.IsNotExist(err) {
```
Comment (ibuildthecloud) [inline main.go]: You are not handling other values of `err` that are not `IsNotExist`

File: main.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+		}
+		fd.Close()
+
+		err = ioutil.WriteFile(hostsWorkingFile, hostData, 0644)
+		if err != nil {
+			return err
+		}
+	}
+
+	hostsFile, err := os.OpenFile(hostsWorkingFile, os.O_RDONLY, 0)
```
Comment (ibuildthecloud) [inline main.go]: `ioutil.ReadFile`

File: main.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	if err != nil {
+		return err
+	}
+	defer hostsFile.Close()
+
+	hostData, err := ioutil.ReadAll(hostsFile)
+
+	lines := strings.Split(string(hostData), "\n")
+
+	for _, line := range lines {
```
Comment (ibuildthecloud) [inline main.go]: What is the advantage of processing this original file content?  Why wouldn't you just copy it byte for byte into the destination?

**Comment by ibuildthecloud** [comment]
Can we rename this repo to `etc-host-updater`, technically we are updating the hostname.

File: main.go (hunk 8 above)
Comment (wlan0) [inline main.go]: @alena1108 and I discussed this, and decided that this service should not overwrite any existing entries in the /etc/hosts file. Therefore, we process the original content to check If we are overwriting. If I copied it byte for byte, then I might have two different entries for the same key.

**Comment by alena1108** [comment]
https://github.com/rancher/rancher/issues/4664

**Pull request merged**